whproxy/proxy.go

```diff
@@ -0,0 +1,214 @@
+package whproxy
+
+import (
+	"bufio"
+	"io"
+	"net/http"
+	"regexp"
+	"sync"
+
+	"github.com/gorilla/websocket"
+	"github.com/taskcluster/webhooktunnel/util"
+	"github.com/taskcluster/webhooktunnel/wsmux"
+)
+
+var (
+	registerRe = regexp.MustCompile("^/register/(\\w+)/?$")
+	serveRe    = regexp.MustCompile("^/(\\w+)/?(.*)$")
+)
+
+// Config for Proxy. Accepts a websocket.Upgrader and a Logger.
+// Default value for Upgrade ReadBufferSize and WriteBufferSize is 1024 bytes.
+// Default Logger is NilLogger.
+type Config struct {
+	Upgrader websocket.Upgrader
+	Logger   util.Logger
+}
+
+// Proxy is used to send http and ws requests to workers.
+// New proxy can be created by using whproxy.New()
+type Proxy struct {
+	m               sync.RWMutex
+	pool            map[string]*wsmux.Session
```

Comment:
Please correct me if I'm wrong, but as far as I can tell, I read through `syncmap` and having `sync.Map` instead of a mutex has the same problem of protecting the data and not the critical section against multiple access.